### PR TITLE
Fix: Password Protection not respected on single product template

### LINKED ISSUE #10939: Password Protection Not Respected on Single Product Template
## Describe the bug

When viewing a password protected Produce on a block-based theme, the password protection is not respected on the frontend and the product is revealed without requiring the password.

On a classic theme, however, the setting is respected.

Using the Classic single product template block for the Single product template it works fine, but as soon as it is converted into blocks the password protection rule is no longer applied on the frontend.

## To reproduce

1. Ensure that you have a product that is password protected. If not, set one to be protected via the **Visibility** setting in the admin (under the **Publish** metabox).
2. Activate the Storefront theme.
3. Visit the password protected product on the frontend.
4. Confirm the product requires the password before showing the product page.
5. Activate the TT3 theme.
6. Visit the password protected product on the frontend.
7. Confirm the product does **_not_** require the password before showing the product page.

## Expected behavior

When a product is protected with a password, the front-end should respect that setting (i.e. it should ask for a password).

## Screenshots

| Classic Theme | Block Theme |
| ------ | ----- |
| CleanShot 2023-09-13 at 17 03 36 | CleanShot 2023-09-13 at 16 59 51 |

### WordPress (please complete the following information):

-   WordPress version: 6.3.1
-   Gutenberg version (if installed): N/A
-   WooCommerce version: 8.1.0
-   WooCommerce Blocks version: 11.2.0-dev
-   Site language: EN
-   Any other plugins installed:

```sh
+-------------------------------+----------+------------------------------+------------+
| name                          | status   | update                       | version    |
+-------------------------------+----------+------------------------------+------------+
| akismet                       | inactive | available                    | 5.2        |
| gutenberg                     | inactive | none                         | 16.6.0     |
| hello                         | inactive | none                         | 1.7.2      |
| jetpack                       | active   | none                         | 12.5       |
| smntcs-show-symlinked-plugins | active   | none                         | 1.1        |
| woocommerce                   | active   | none                         | 8.1.0      |
| woocommerce-blocks            | active   | version higher than expected | 11.2.0-dev |
| wordpress-importer            | active   | none                         | 0.8.1      |
+-------------------------------+----------+------------------------------+------------+
```

### Desktop (please complete the following information):

-   OS: macOS
-   Browser: Chrome
-   Version: 116.0.5845.179

### Smartphone (please complete the following information):

-   Device: [e.g. iPhone6]
-   OS: [e.g. iOS8.1]
-   Browser [e.g. stock browser, safari]
-   Version [e.g. 22]

## Additional context

- Related issue in WC Core: https://github.com/woocommerce/woocommerce/issues/39648
- Slack conversation: p1694633746333029-slack-C02FL3X7KR6

### PR DESCRIPTION
**Pull request by gigitux**
## What

Fixes #10939 

## Why

The function `add_password_form` adds the password form to the single product template when the first block that belongs to the single product block is founded. The rest of the blocks related to the Single Product template are removed.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Ensure you are using the Blockified Single Product Template.
2. Create a product with password protection.
3. Visit the Product.
4. Ensure that the password form is added and the other blocks related to the Single Product Template are not visible.

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

image

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [ ] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Fix: Implement password protection on the blockified Single Product Template.

### CODE REVIEW

File: src/Templates/SingleProductTemplate.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Templates;
+
+/**
+ * SingleProductTemplae class.
+ *
+ * @internal
+ */
+class SingleProductTemplate {
+
```
Comment (danieldudzic) [inline src/Templates/SingleProductTemplate.php:10]: Are two empty lines necessary?
```suggestion

```

**Review by danieldudzic — APPROVED** [review]
Really, really solid work on this @gigitux! Tough problem, but I believe you have solved it very gracefully.


I was able to replicate the original issue with `trunk` and this PR indeed fixes it ✅

File: src/BlockTemplatesController.php [hunk 2]
```diff
@@ -429,8 +430,12 @@ function( $classes ) {
 							}
 						}
 
-						$new_content       = SingleProductTemplateCompatibility::add_compatibility_layer( $template->content );
-						$template->content = $new_content;
+						if ( post_password_required() ) {
+							$template->content = SingleProductTemplate::add_password_form( $template->content );
+						} else {
+							$new_content       = SingleProductTemplateCompatibility::add_compatibility_layer( $template->content );
+							$template->content = $new_content;
```
Comment (tjcafferkey) [inline src/BlockTemplatesController.php:437]: Could we not have just done:

```php
$template->content = SingleProductTemplateCompatibility::add_compatibility_layer( $template->content );
```

**Comment by tjcafferkey** [comment]
Appreciate the tests that have come with this also @gigitux. Thank you!

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (gigitux) [reply src/BlockTemplatesController.php:437]: True! I didn't change that code, but I just moved! Should I create a follow-up PR?

Comment (tjcafferkey) [reply src/BlockTemplatesController.php:437]: I don't think it's urgent. I think I'll create a follow up for it later and you can review if that works 😃

**Pull request merged**